lib/handlers/events/delivery/goals/k8s/KubernetesGoalScheduler.ts

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    automationClientInstance,
+    AutomationContextAware,
+    configurationValue,
+    doWithRetry,
+    HandlerContext,
+    logger,
+} from "@atomist/automation-client";
+import {
+    ExecuteGoalResult,
+    GoalInvocation,
+    GoalScheduler,
+    SdmGoalEvent,
+} from "@atomist/sdm";
+import * as k8s from "@kubernetes/client-node";
+import * as cluster from "cluster";
+import * as _ from "lodash";
+import * as os from "os";
+import { loadKubeConfig } from "./k8config";
+
+/**
+ * GoalScheduler implementation that schedules SDM goals inside k8s jobs.
+ *
+ * It reuses the podSpec of the deployed SDM to create a new jobSpec from.
+ * Subclasses may change the spec and job creation behavior by overwriting beforeCreation
+ * and/or afterCreation methods.
+ */
+export class KubernetesGoalScheduler implements GoalScheduler {
+
+    constructor(private readonly options: { isolateAll?: boolean } = { isolateAll: false }) {
+        this.init();
+    }
+
+    public async supports(gi: GoalInvocation): Promise<boolean> {
+        return !process.env.ATOMIST_ISOLATED_GOAL &&
+            (
+                // Goal is marked as isolated and SDM is configured to use k8s jobs
+                (gi.goal.definition.isolated && isConfiguredInEnv("kubernetes")) ||
+                // Force all goals to run isolated via env var
+                isConfiguredInEnv("kubernetes-all") ||
+                // Force all goals to run isolated via explicit option
+                (this.options.isolateAll && isConfiguredInEnv("kubernetes")) ||
+                // Force all goals to run isolated via explicit configuration
+                _.get(gi.configuration, "sdm.kubernetes.isolateAll", false) === true
```

Comment:
This is a bit of a conundrum. `sdm.kubernetes.isolateAll` isn't right, nor is `sdm.goal.isolateAll`. Maybe `isolateAll` should be a string and its value should be "kubernetes".